File: c++/task2_employees.cpp
#include "task2_employees.h"
#include "benchmark_utils.h"
#include <iostream>
#include <thread>
#include <vector>
#include <random>
#include <algorithm>
#include <mutex>
#include <iomanip>
#include <cmath>
#include <sstream>

namespace task2 {

std::vector<Employee> generate_employees(int count, const std::string& target_position) {
    std::vector<Employee> employees;
    std::random_device rd;
    std::mt19937 gen(rd());
    
    // Списки для генерации данных
    std::vector<std::string> first_names = {"Иван", "Петр", "Сергей", "Алексей", "Дмитрий", 
                                           "Мария", "Ольга", "Елена", "Анна", "Наталья"};
    std::vector<std::string> last_names = {"Иванов", "Петров", "Сидоров", "Смирнов", "Кузнецов",
                                          "Попов", "Васильев", "Павлов", "Семенов", "Федоров"};
    std::vector<std::string> middle_names = {"Иванович", "Петрович", "Сергеевич", "Алексеевич", 
                                            "Дмитриевич", "Ивановна", "Петровна", "Сергеевна", 
                                            "Алексеевна", "Дмитриевна"};
    
    std::vector<std::string> positions = {"Менеджер", "Разработчик", "Аналитик", "Тестировщик", 
                                         "Дизайнер", "Администратор", "Бухгалтер", target_position};
    
    std::uniform_int_distribution<> age_dist(20, 65);
    std::uniform_real_distribution<> salary_dist(30000, 300000);
    std::uniform_int_distribution<> position_dist(0, positions.size() - 1);
    
    for (int i = 0; i < count; ++i) {
        // Генерация ФИО
        std::ostringstream name;
        name << last_names[gen() % last_names.size()] << " "
             << first_names[gen() % first_names.size()] << " "
             << middle_names[gen() % middle_names.size()];
        
        // Генерация должности
        std::string position = positions[position_dist(gen)];
        
        // Генерация возраста
        int age = age_dist(gen);
        
        // Генерация зарплаты
        double salary = salary_dist(gen);
        
        employees.emplace_back(name.str(), position, age, salary);
    }
    
    // Убедимся, что есть сотрудники с целевой должностью
    bool has_target_position = false;
    for (const auto& emp : employees) {
        if (emp.position == target_position) {
            has_target_position = true;
            break;
        }
    }
    
    if (!has_target_position && !employees.empty()) {
        employees[0].position = target_position;
    }
    
    return employees;
}

double calculate_average_age(const std::vector<Employee>& employees, const std::string& target_position) {
    double total_age = 0.0;
    int count = 0;
    
    for (const auto& emp : employees) {
        if (emp.position == target_position) {
            total_age += emp.age;
            count++;
        }
    }
    
    return count > 0 ? total_age / count : 0.0;
}

double find_max_salary_near_average(const std::vector<Employee>& employees, 
                                   const std::string& target_position, 
                                   double average_age, 
                                   int age_range) {
    double max_salary = 0.0;
    
    for (const auto& emp : employees) {
        if (emp.position == target_position && 
            std::abs(emp.age - average_age) <= age_range) {
            if (emp.salary > max_salary) {
                max_salary = emp.salary;
            }
        }
    }
    
    return max_salary;
}

void process_single_thread(const std::vector<Employee>& employees, 
                          const std::string& target_position) {
    // Расчет среднего возраста
    double average_age = calculate_average_age(employees, target_position);
    
    // Поиск максимальной зарплаты
    double max_salary = find_max_salary_near_average(employees, target_position, average_age);
    
    // Подсчет количества сотрудников с целевой должностью
    int target_count = 0;
    for (const auto& emp : employees) {
        if (emp.position == target_position) {
            target_count++;
        }
    }
    
    std::cout << "\n=== Результаты обработки (однопоточная) ===\n";
    std::cout << "Всего сотрудников: " << employees.size() << "\n";
    std::cout << "Сотрудников с должностью '" << target_position << "': " << target_count << "\n\n";
    
    if (target_count > 0) {
        std::cout << "Средний возраст: " << std::fixed << std::setprecision(2) << average_age << " лет\n";
        std::cout << "Максимальная зарплата среди сотрудников\n";
        std::cout << "с возрастом ±2 года от среднего: " 
                  << std::fixed << std::setprecision(2) << max_salary << " руб.\n";
    } else {
        std::cout << "Нет сотрудников с должностью '" << target_position << "'\n";
    }
}

void process_multi_thread(const std::vector<Employee>& employees, 
                         const std::string& target_position, 
                         int num_threads) {
    if (employees.empty()) {
        std::cout << "Нет данных для обработки\n";
        return;
    }
    
    std::vector<std::thread> threads;
    std::vector<double> thread_ages(num_threads, 0.0);
    std::vector<int> thread_counts(num_threads, 0);
    std::vector<double> thread_max_salaries(num_threads, 0.0);
    
    int chunk_size = employees.size() / num_threads;
    std::mutex cout_mutex;
    
    for (int i = 0; i < num_threads; ++i) {
        threads.emplace_back([&, i]() {
            int start = i * chunk_size;
            int end = (i == num_threads - 1) ? employees.size() : start + chunk_size;
            
            for (int j = start; j < end; ++j) {
                const auto& emp = employees[j];
                if (emp.position == target_position) {
                    thread_ages[i] += emp.age;
                    thread_counts[i]++;
                    
                    // Пока не знаем средний возраст, сохраняем максимальную зарплату
                    if (emp.salary > thread_max_salaries[i]) {
                        thread_max_salaries[i] = emp.salary;
                    }
                }
            }
        });
    }
    
    for (auto& t : threads) {
        t.join();
    }
    
    // Агрегируем результаты
    double total_age = 0.0;
    int total_count = 0;
    
    for (int i = 0; i < num_threads; ++i) {
        total_age += thread_ages[i];
        total_count += thread_counts[i];
    }
    
    double average_age = total_count > 0 ? total_age / total_count : 0.0;
    
    // Вторая фаза: поиск максимальной зарплаты с учетом среднего возраста
    threads.clear();
    std::vector<double> thread_phase2_max(num_threads, 0.0);
    
    for (int i = 0; i < num_threads; ++i) {
        threads.emplace_back([&, i, average_age]() {
            int start = i * chunk_size;
            int end = (i == num_threads - 1) ? employees.size() : start + chunk_size;
            
            for (int j = start; j < end; ++j) {
                const auto& emp = employees[j];
                if (emp.position == target_position && 
                    std::abs(emp.age - average_age) <= 2) {
                    if (emp.salary > thread_phase2_max[i]) {
                        thread_phase2_max[i] = emp.salary;
                    }
                }
            }
        });
    }
    
    for (auto& t : threads) {
        t.join();
    }
    
    double max_salary = 0.0;
    for (int i = 0; i < num_threads; ++i) {
        if (thread_phase2_max[i] > max_salary) {
            max_salary = thread_phase2_max[i];
        }
    }
    
    std::cout << "\n=== Результаты обработки (многопоточная) ===\n";
    std::cout << "Использовано потоков: " << num_threads << "\n";
    std::cout << "Всего сотрудников: " << employees.size() << "\n";
    std::cout << "Сотрудников с должностью '" << target_position << "': " << total_count << "\n\n";
    
    if (total_count > 0) {
        std::cout << "Средний возраст: " << std::fixed << std::setprecision(2) << average_age << " лет\n";
        std::cout << "Максимальная зарплата среди сотрудников\n";
        std::cout << "с возрастом ±2 года от среднего: " 
                  << std::fixed << std::setprecision(2) << max_salary << " руб.\n";
    } else {
        std::cout << "Нет сотрудников с должностью '" << target_position << "'\n";
    }
}

void analyze_performance(int min_size, int max_size, int step, 
                        const std::string& target_position) {
    std::cout << "\n=== Анализ производительности ===\n";
    std::cout << "Тестируем обработку разных объемов данных\n";
    std::cout << "Целевая должность: '" << target_position << "'\n\n";
    
    std::vector<std::pair<std::string, double>> single_thread_results;
    std::vector<std::pair<std::string, double>> multi_thread_results;
    
    for (int size = min_size; size <= max_size; size += step) {
        std::cout << "Тест с " << size << " сотрудниками...\n";
        
        auto employees = generate_employees(size, target_position);
        
        double single_time, multi_time;
        
        {
            Benchmark b("Однопоточная", false);
            process_single_thread(employees, target_position);
            single_time = b.elapsed_microseconds();
        }
        
        {
            Benchmark b("Многопоточная (4 потока)", false);
            process_multi_thread(employees, target_position, 4);
            multi_time = b.elapsed_microseconds();
        }
        
        single_thread_results.emplace_back(std::to_string(size), single_time);
        multi_thread_results.emplace_back(std::to_string(size), multi_time);
        
        double speedup = single_time / multi_time;
        std::cout << "  Ускорение: " << std::fixed << std::setprecision(2) << speedup << "x\n\n";
    }
    
    std::cout << "\n=== Итоги анализа производительности ===\n";
    std::cout << std::setw(10) << "Размер" 
              << std::setw(20) << "Однопоточная (мс)"
              << std::setw(20) << "Многопоточная (мс)"
              << std::setw(15) << "Ускорение" << "\n";
    std::cout << std::string(65, '-') << std::endl;
    
    for (size_t i = 0; i < single_thread_results.size(); ++i) {
        double single_ms = single_thread_results[i].second / 1000.0;
        double multi_ms = multi_thread_results[i].second / 1000.0;
        double speedup = single_thread_results[i].second / multi_thread_results[i].second;
        
        std::cout << std::setw(10) << single_thread_results[i].first
                  << std::setw(20) << std::fixed << std::setprecision(2) << single_ms
                  << std::setw(20) << std::fixed << std::setprecision(2) << multi_ms
                  << std::setw(15) << std::fixed << std::setprecision(2) << speedup << "x\n";
    }
    std::cout << std::string(65, '-') << std::endl;
}

void run_employees_benchmark() {
    std::cout << "\n=== Бенчмарк анализа сотрудников (вариант 26) ===\n";
    
    std::string target_position = "Инженер";
    std::vector<int> test_sizes = {1000, 5000, 10000, 50000, 100000};
    std::vector<int> thread_counts = {1, 2, 4, 8};
    
    std::vector<std::pair<std::string, double>> benchmark_results;
    
    for (int size : test_sizes) {
        std::cout << "\nГенерация " << size << " сотрудников...\n";
        auto employees = generate_employees(size, target_position);
        
        for (int threads : thread_counts) {
            std::string test_name = std::to_string(size) + "_сотр_" + std::to_string(threads) + "_потоков";
            
            Benchmark b(test_name, false);
            if (threads == 1) {
                process_single_thread(employees, target_position);
            } else {
                process_multi_thread(employees, target_position, threads);
            }
            
            benchmark_results.emplace_back(test_name, b.elapsed_microseconds());
        }
    }
    
    Benchmark::save_to_csv(benchmark_results, "employees_benchmark.csv");
    std::cout << "\nБенчмарк завершен. Результаты сохранены в employees_benchmark.csv\n";
}

void run_employees() {
    std::cout << "\n=== Задание 2: Анализ сотрудников (вариант 26) ===\n";
    std::cout << "Найти средний возраст для должности Д\n";
    std::cout << "и наибольшую зарплату среди сотрудников должности Д,\n";
    std::cout << "чья возраст отличается от среднего не более чем на 2 года.\n\n";
    
    int choice;
    std::cout << "Выберите режим:\n";
    std::cout << "1. Стандартный анализ\n";
    std::cout << "2. Анализ производительности\n";
    std::cout << "3. Полный бенчмарк\n";
    std::cout << "Ваш выбор: ";
    std::cin >> choice;
    
    std::string target_position;
    std::cout << "\nВведите целевую должность (Д): ";
    std::cin.ignore();
    std::getline(std::cin, target_position);
    
    if (target_position.empty()) {
        target_position = "Инженер";
    }
    
    switch (choice) {
        case 1: {
            int num_employees, num_threads;
            
            std::cout << "\nВведите количество сотрудников (100-100000): ";
            std::cin >> num_employees;
            
            if (num_employees < 100) num_employees = 100;
            if (num_employees > 100000) num_employees = 100000;
            
            std::cout << "Генерация " << num_employees << " сотрудников...\n";
            auto employees = generate_employees(num_employees, target_position);
            
            double single_time, multi_time;
            
            {
                Benchmark b("Однопоточная обработка");
                process_single_thread(employees, target_position);
                single_time = b.elapsed_microseconds();
            }
            
            std::cout << "\nВведите количество потоков для многопоточной обработки (2-16): ";
            std::cin >> num_threads;
            
            if (num_threads < 2) num_threads = 2;
            if (num_threads > 16) num_threads = 16;
            
            {
                Benchmark b("Многопоточная обработка");
                process_multi_thread(employees, target_position, num_threads);
                multi_time = b.elapsed_microseconds();
            }
            
            Benchmark::print_comparison("Однопоточная", single_time, 
                                       "Многопоточная (" + std::to_string(num_threads) + " потоков)", 
                                       multi_time);
            break;
        }
        case 2:
            analyze_performance(1000, 10000, 2000, target_position);
            break;
        case 3:
            run_employees_benchmark();
            break;
        default:
            std::cout << "Неверный выбор! Запускаю стандартный анализ...\n";
            auto employees = generate_employees(5000, target_position);
            process_single_thread(employees, target_position);
            process_multi_thread(employees, target_position, 4);
    }
}

} // namespace task2

File: go/task2_employees.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

// Employee представляет сотрудника
type Employee struct {
	Name     string  // ФИО
	Position string  // Должность
	Age      int     // Возраст
	Salary   float64 // Заработная плата
}

// GenerateEmployees генерирует список сотрудников
func GenerateEmployees(count int, targetPosition string) []Employee {
	employees := make([]Employee, 0, count)

	// Списки для генерации данных
	firstNames := []string{"Иван", "Петр", "Сергей", "Алексей", "Дмитрий",
		"Мария", "Ольга", "Елена", "Анна", "Наталья"}
	lastNames := []string{"Иванов", "Петров", "Сидоров", "Смирнов", "Кузнецов",
		"Попов", "Васильев", "Павлов", "Семенов", "Федоров"}
	middleNames := []string{"Иванович", "Петрович", "Сергеевич", "Алексеевич",
		"Дмитриевич", "Ивановна", "Петровна", "Сергеевна", "Алексеевна", "Дмитриевна"}

	positions := []string{"Менеджер", "Разработчик", "Аналитик", "Тестировщик",
		"Дизайнер", "Администратор", "Бухгалтер", targetPosition}

	rng := rand.New(rand.NewSource(time.Now().UnixNano()))

	for i := 0; i < count; i++ {
		// Генерация ФИО
		name := lastNames[rng.Intn(len(lastNames))] + " " +
			firstNames[rng.Intn(len(firstNames))] + " " +
			middleNames[rng.Intn(len(middleNames))]

		// Генерация должности
		position := positions[rng.Intn(len(positions))]

		// Генерация возраста (20-65 лет)
		age := 20 + rng.Intn(46)

		// Генерация зарплаты (30,000 - 300,000 руб.)
		salary := 30000 + rng.Float64()*270000

		employees = append(employees, Employee{
			Name:     name,
			Position: position,
			Age:      age,
			Salary:   salary,
		})
	}

	// Убедимся, что есть сотрудники с целевой должностью
	hasTargetPosition := false
	for _, emp := range employees {
		if emp.Position == targetPosition {
			hasTargetPosition = true
			break
		}
	}

	if !hasTargetPosition && len(employees) > 0 {
		employees[0].Position = targetPosition
	}

	return employees
}

// CalculateAverageAge вычисляет средний возраст для указанной должности
func CalculateAverageAge(employees []Employee, targetPosition string) float64 {
	var totalAge int
	var count int

	for _, emp := range employees {
		if emp.Position == targetPosition {
			totalAge += emp.Age
			count++
		}
	}

	if count == 0 {
		return 0
	}
	return float64(totalAge) / float64(count)
}

// FindMaxSalaryNearAverage находит максимальную зарплату среди сотрудников,
// чей возраст отличается от среднего не более чем на ageRange лет
func FindMaxSalaryNearAverage(employees []Employee, targetPosition string,
	averageAge float64, ageRange int) float64 {
	maxSalary := 0.0

	for _, emp := range employees {
		if emp.Position == targetPosition {
			ageDiff := float64(emp.Age) - averageAge
			if ageDiff < 0 {
				ageDiff = -ageDiff
			}
			if ageDiff <= float64(ageRange) {
				if emp.Salary > maxSalary {
					maxSalary = emp.Salary
				}
			}
		}
	}

	return maxSalary
}

// ProcessSingleThread обрабатывает сотрудников в одном потоке
func ProcessSingleThread(employees []Employee, targetPosition string) {
	// Расчет среднего возраста
	averageAge := CalculateAverageAge(employees, targetPosition)

	// Поиск максимальной зарплаты
	maxSalary := FindMaxSalaryNearAverage(employees, targetPosition, averageAge, 2)

	// Подсчет количества сотрудников с целевой должности
	targetCount := 0
	for _, emp := range employees {
		if emp.Position == targetPosition {
			targetCount++
		}
	}

	fmt.Println("\n=== Результаты обработки (однопоточная) ===")
	fmt.Printf("Всего сотрудников: %d\n", len(employees))
	fmt.Printf("Сотрудников с должностью '%s': %d\n\n", targetPosition, targetCount)

	if targetCount > 0 {
		fmt.Printf("Средний возраст: %.2f лет\n", averageAge)
		fmt.Println("Максимальная зарплата среди сотрудников")
		fmt.Printf("с возрастом ±2 года от среднего: %.2f руб.\n", maxSalary)
	} else {
		fmt.Printf("Нет сотрудников с должностью '%s'\n", targetPosition)
	}
}

// ProcessMultiThread обрабатывает сотрудников в нескольких потоках
func ProcessMultiThread(employees []Employee, targetPosition string, numThreads int) {
	if len(employees) == 0 {
		fmt.Println("Нет данных для обработки")
		return
	}

	var wg sync.WaitGroup
	threadAges := make([]float64, numThreads)
	threadCounts := make([]int, numThreads)
	threadMaxSalaries := make([]float64, numThreads)

	chunkSize := len(employees) / numThreads

	// Первая фаза: вычисление среднего возраста
	for i := 0; i < numThreads; i++ {
		wg.Add(1)
		go func(threadID int) {
			defer wg.Done()

			start := threadID * chunkSize
			end := start + chunkSize
			if threadID == numThreads-1 {
				end = len(employees)
			}

			for j := start; j < end; j++ {
				emp := employees[j]
				if emp.Position == targetPosition {
					threadAges[threadID] += float64(emp.Age)
					threadCounts[threadID]++

					// Сохраняем максимальную зарплату (пока не знаем средний возраст)
					if emp.Salary > threadMaxSalaries[threadID] {
						threadMaxSalaries[threadID] = emp.Salary
					}
				}
			}
		}(i)
	}

	wg.Wait()

	// Агрегируем результаты для расчета среднего возраста
	var totalAge float64
	var totalCount int
	for i := 0; i < numThreads; i++ {
		totalAge += threadAges[i]
		totalCount += threadCounts[i]
	}

	averageAge := 0.0
	if totalCount > 0 {
		averageAge = totalAge / float64(totalCount)
	}

	// Вторая фаза: поиск максимальной зарплаты с учетом среднего возраста
	wg = sync.WaitGroup{}
	threadPhase2Max := make([]float64, numThreads)

	for i := 0; i < numThreads; i++ {
		wg.Add(1)
		go func(threadID int, avgAge float64) {
			defer wg.Done()

			start := threadID * chunkSize
			end := start + chunkSize
			if threadID == numThreads-1 {
				end = len(employees)
			}

			for j := start; j < end; j++ {
				emp := employees[j]
				if emp.Position == targetPosition {
					ageDiff := float64(emp.Age) - avgAge
					if ageDiff < 0 {
						ageDiff = -ageDiff
					}
					if ageDiff <= 2 {
						if emp.Salary > threadPhase2Max[threadID] {
							threadPhase2Max[threadID] = emp.Salary
						}
					}
				}
			}
		}(i, averageAge)
	}

	wg.Wait()

	// Находим общую максимальную зарплату
	maxSalary := 0.0
	for i := 0; i < numThreads; i++ {
		if threadPhase2Max[i] > maxSalary {
			maxSalary = threadPhase2Max[i]
		}
	}

	fmt.Println("\n=== Результаты обработки (многопоточная) ===")
	fmt.Printf("Использовано потоков: %d\n", numThreads)
	fmt.Printf("Всего сотрудников: %d\n", len(employees))
	fmt.Printf("Сотрудников с должностью '%s': %d\n\n", targetPosition, totalCount)

	if totalCount > 0 {
		fmt.Printf("Средний возраст: %.2f лет\n", averageAge)
		fmt.Println("Максимальная зарплата среди сотрудников")
		fmt.Printf("с возрастом ±2 года от среднего: %.2f руб.\n", maxSalary)
	} else {
		fmt.Printf("Нет сотрудников с должностью '%s'\n", targetPosition)
	}
}

// RunEmployees запускает задание 2
func RunEmployees() {
	fmt.Println("\n=== Задание 2: Анализ сотрудников (вариант 26) ===")
	fmt.Println("Найти средний возраст для должности Д")
	fmt.Println("и наибольшую зарплату среди сотрудников должности Д,")
	fmt.Println("чья возраст отличается от среднего не более чем на 2 года.\n")

	fmt.Print("Введите целевую должность (Д): ")
	var targetPosition string
	fmt.Scanln()
	fmt.Scanln(&targetPosition)

	if targetPosition == "" {
		targetPosition = "Инженер"
	}

	var choice string
	fmt.Println("\nВыберите режим:")
	fmt.Println("1. Стандартный анализ")
	fmt.Println("2. Анализ производительности")
	fmt.Println("3. Полный бенчмарк")
	fmt.Print("Ваш выбор: ")
	fmt.Scan(&choice)

	switch choice {
	case "1":
		var numEmployees, numThreads int

		fmt.Print("\nВведите количество сотрудников (100-100000): ")
		fmt.Scan(&numEmployees)

		if numEmployees < 100 {
			numEmployees = 100
		}
		if numEmployees > 100000 {
			numEmployees = 100000
		}

		fmt.Printf("Генерация %d сотрудников...\n", numEmployees)
		employees := GenerateEmployees(numEmployees, targetPosition)

		var singleTime, multiTime float64

		// Однопоточная обработка
		{
			b := NewBenchmark("Однопоточная обработка")
			ProcessSingleThread(employees, targetPosition)
			singleTime = b.StopSilent()
		}

		fmt.Print("\nВведите количество потоков для многопоточной обработки (2-16): ")
		fmt.Scan(&numThreads)

		if numThreads < 2 {
			numThreads = 2
		}
		if numThreads > 16 {
			numThreads = 16
		}

		// Многопоточная обработка
		{
			b := NewBenchmark("Многопоточная обработка")
			ProcessMultiThread(employees, targetPosition, numThreads)
			multiTime = b.StopSilent()
		}

		PrintComparison("Однопоточная", singleTime,
			fmt.Sprintf("Многопоточная (%d потоков)", numThreads), multiTime)

	case "2":
		AnalyzePerformance(1000, 10000, 2000, targetPosition)

	case "3":
		RunEmployeesBenchmark()

	default:
		fmt.Println("Неверный выбор! Запускаю стандартный анализ...")
		employees := GenerateEmployees(5000, targetPosition)
		ProcessSingleThread(employees, targetPosition)
		ProcessMultiThread(employees, targetPosition, 4)
	}
}

// RunEmployeesBenchmark запускает бенчмарк анализа сотрудников
func RunEmployeesBenchmark() {
	fmt.Println("\n=== Бенчмарк анализа сотрудников (вариант 26) ===")

	fmt.Print("Введите целевую должность (по умолчанию 'Инженер'): ")
	var targetPosition string
	fmt.Scanln()
	fmt.Scanln(&targetPosition)
	if targetPosition == "" {
		targetPosition = "Инженер"
	}

	testSizes := []int{1000, 5000, 10000, 50000, 100000}
	threadCounts := []int{1, 2, 4, 8}

	benchmarkResults := make([]BenchmarkResult, 0)

	for _, size := range testSizes {
		fmt.Printf("\nГенерация %d сотрудников...\n", size)
		employees := GenerateEmployees(size, targetPosition)

		for _, threads := range threadCounts {
			testName := fmt.Sprintf("%d_сотр_%d_потоков_%s", size, threads, targetPosition)

			b := NewBenchmark(testName, false)

			if threads == 1 {
				ProcessSingleThread(employees, targetPosition)
			} else {
				ProcessMultiThread(employees, targetPosition, threads)
			}

			time := b.StopSilent()
			benchmarkResults = append(benchmarkResults, BenchmarkResult{
				Name:             testName,
				TimeMicroseconds: time,
				TimeMilliseconds: time / 1000.0,
				TimeSeconds:      time / 1000000.0,
			})
		}
	}

	SaveToCSV(benchmarkResults, "employees_benchmark.csv")
	fmt.Println("\nБенчмарк завершен. Результаты сохранены в employees_benchmark.csv")
}

// AnalyzePerformance анализирует производительность
func AnalyzePerformance(minSize, maxSize, step int, targetPosition string) {
	fmt.Println("\n=== Анализ производительности ===")
	fmt.Printf("Целевая должность: '%s'\n\n", targetPosition)

	singleThreadResults := make([]BenchmarkResult, 0)
	multiThreadResults := make([]BenchmarkResult, 0)

	for size := minSize; size <= maxSize; size += step {
		fmt.Printf("Тест с %d сотрудниками...\n", size)

		employees := GenerateEmployees(size, targetPosition)

		var singleTime, multiTime float64

		// Однопоточная обработка
		{
			b := NewBenchmark("Однопоточная", false)
			ProcessSingleThread(employees, targetPosition)
			singleTime = b.StopSilent()
		}

		// Многопоточная обработка (4 потока)
		{
			b := NewBenchmark("Многопоточная (4 потока)", false)
			ProcessMultiThread(employees, targetPosition, 4)
			multiTime = b.StopSilent()
		}

		singleThreadResults = append(singleThreadResults, BenchmarkResult{
			Name:             fmt.Sprintf("%d", size),
			TimeMicroseconds: singleTime,
			TimeMilliseconds: singleTime / 1000.0,
			TimeSeconds:      singleTime / 1000000.0,
		})

		multiThreadResults = append(multiThreadResults, BenchmarkResult{
			Name:             fmt.Sprintf("%d", size),
			TimeMicroseconds: multiTime,
			TimeMilliseconds: multiTime / 1000.0,
			TimeSeconds:      multiTime / 1000000.0,
		})

		speedup := singleTime / multiTime
		fmt.Printf("  Ускорение: %.2fx\n\n", speedup)
	}

	fmt.Println("\n=== Итоги анализа производительности ===")
	fmt.Printf("%-10s %-20s %-20s %-15s\n", "Размер", "Однопоточная (мс)", "Многопоточная (мс)", "Ускорение")
	fmt.Println("-----------------------------------------------------------------")

	for i := 0; i < len(singleThreadResults); i++ {
		singleMS := singleThreadResults[i].TimeMilliseconds
		multiMS := multiThreadResults[i].TimeMilliseconds
		speedup := singleThreadResults[i].TimeMicroseconds / multiThreadResults[i].TimeMicroseconds

		fmt.Printf("%-10s %-20.2f %-20.2f %-15.2fx\n",
			singleThreadResults[i].Name, singleMS, multiMS, speedup)
	}
	fmt.Println("-----------------------------------------------------------------")
}
